### ISSUE TITLE
Hide inline basic auth part of remoteWrite.url in logs of vmagent

### ISSUE THREAD

**Issue by ofen** [VictoriaMetrics/VictoriaMetrics#773]
**Describe the solution you'd like**
Hide password in logs in case if user want's to use basic auth credentials as part of remoteWrite.url e.g. `--remoteWrite.url=https://username:my_password@remote-storage.com/api/v1/write`

**Additional context**
If username and password is set this way they will be showed up in logs and will be **partially masked** in error logs:
```
2020-09-16T15:54:37.079Z    info    VictoriaMetrics/lib/logger/flag.go:11   build version: vmagent-20200815-125807-tags-v1.40.0-0-ged00eb3f3
2020-09-16T15:54:37.079Z    info    VictoriaMetrics/lib/logger/flag.go:12   command line flags
2020-09-16T15:54:37.079Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "csvTrimTimestamp" = "1ms"
2020-09-16T15:54:37.079Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "dryRun" = "false"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "enableTCP6" = "false"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "envflag.enable" = "false"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "envflag.prefix" = ""
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "fs.disableMmap" = "false"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "graphiteListenAddr" = ""
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "graphiteTrimTimestamp" = "1s"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "http.disableResponseCompression" = "false"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "http.maxGracefulShutdownDuration" = "7s"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "http.pathPrefix" = ""
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "http.shutdownDelay" = "0s"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "httpAuth.password" = "secret"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "httpAuth.username" = ""
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "httpListenAddr" = ":8429"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "import.maxLineLen" = "104857600"
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "influxListenAddr" = ""
2020-09-16T15:54:37.080Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "influxMeasurementFieldSeparator" = "_"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "influxSkipMeasurement" = "false"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "influxSkipSingleField" = "false"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "influxTrimTimestamp" = "1ms"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "insert.maxQueueDuration" = "1m0s"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "loggerErrorsPerSecondLimit" = "10"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "loggerFormat" = "default"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "loggerLevel" = "INFO"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "loggerOutput" = "stderr"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "maxConcurrentInserts" = "16"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "maxInsertRequestSize" = "33554432"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "memory.allowedBytes" = "0"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "memory.allowedPercent" = "60"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "metricsAuthKey" = "secret"
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "opentsdbHTTPListenAddr" = ""
2020-09-16T15:54:37.081Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "opentsdbListenAddr" = ""
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "opentsdbTrimTimestamp" = "1s"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "opentsdbhttp.maxInsertRequestSize" = "33554432"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "opentsdbhttpTrimTimestamp" = "1ms"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "pprofAuthKey" = "secret"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.config" = "/config/agent.yaml"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.config.dryRun" = "false"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.config.strictParse" = "true"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.configCheckInterval" = "0s"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.consulSDCheckInterval" = "30s"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.disableCompression" = "false"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.disableKeepAlive" = "false"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.discovery.concurrency" = "100"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.discovery.concurrentWaitTime" = "1m0s"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.dnsSDCheckInterval" = "30s"
2020-09-16T15:54:37.082Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.ec2SDCheckInterval" = "1m0s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.fileSDCheckInterval" = "30s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.gceSDCheckInterval" = "1m0s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.kubernetesSDCheckInterval" = "30s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.maxScrapeSize" = "16777216"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "promscrape.suppressScrapeErrors" = "false"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.basicAuth.password" = "secret"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.basicAuth.username" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.bearerToken" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.decimalPlaces" = "0"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.flushInterval" = "1s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.label" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.maxBlockSize" = "33554432"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.maxDiskUsagePerURL" = "0"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.proxyURL" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.queues" = "20"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.relabelConfig" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.sendTimeout" = "1m0s"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.showURL" = "false"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tlsCAFile" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tlsCertFile" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tlsInsecureSkipVerify" = "false"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tlsKeyFile" = "secret"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tlsServerName" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.tmpDataPath" = "vmagent-remotewrite-data"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.url" = "https://agent:my_password@remote-storage.com/api/v1/write"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "remoteWrite.urlRelabelConfig" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "tls" = "false"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "tlsCertFile" = ""
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "tlsKeyFile" = "secret"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/logger/flag.go:20   flag "version" = "false"
2020-09-16T15:54:37.083Z    info    VictoriaMetrics/lib/cgroup/cpu.go:26    updating GOMAXPROCS to 1 according to cgroup CPU quota
2020-09-16T15:54:37.084Z    info    VictoriaMetrics/app/vmagent/main.go:80  starting vmagent at ":8429"...
2020-09-16T15:54:37.084Z    info    VictoriaMetrics/lib/memory/memory.go:45 limiting caches to 2576980377 bytes, leaving 1717986919 bytes to the OS according to -memory.allowedPercent=60.000000 and -memory.allowedBytes=0
2020-09-16T15:54:37.087Z    info    VictoriaMetrics/lib/persistentqueue/fastqueue.go:46 opened fast persistent queue at "vmagent-remotewrite-data/persistent-queue/AF03C634A296195F" with maxInmemoryBlocks=200
2020-09-16T15:54:37.087Z    info    VictoriaMetrics/app/vmagent/remotewrite/client.go:125   initialized client for -remoteWrite.url="https://agent:my_password@remote-storage.com/api/v1/write"
2020-09-16T15:54:37.087Z    info    VictoriaMetrics/app/vmagent/main.go:102 started vmagent in 0.004 seconds
2020-09-16T15:54:37.087Z    info    VictoriaMetrics/lib/httpserver/httpserver.go:77 starting http server at http://:8429/
2020-09-16T15:54:37.087Z    info    VictoriaMetrics/lib/httpserver/httpserver.go:78 pprof handlers are exposed at http://:8429/debug/pprof/
2020-09-16T15:54:37.088Z    info    VictoriaMetrics/lib/promscrape/scraper.go:78    reading Prometheus configs from "/config/agent.yaml"
2020-09-16T15:54:37.093Z    info    VictoriaMetrics/lib/promscrape/scraper.go:303   static_configs: added targets: 1, removed targets: 0; total targets: 1
2020-09-16T15:54:37.363Z    info    VictoriaMetrics/lib/promscrape/scraper.go:303   kubernetes_sd_configs: added targets: 62, removed targets: 0; total targets: 62
2020-09-16T16:30:29.135Z    error   VictoriaMetrics/app/vmagent/remotewrite/client.go:212   couldn't send a block with size 151051 bytes to "https://agent:my_password@remote-storage.com/api/v1/write": Post "https://username:***@remote-storage.com/api/v1/write": EOF; re-sending the block in 2.000 seconds
```

**Comment by valyala** [comment]
Note that basic auth credentials may be passed to `vmagent` via `-remoteWrite.basicAuth.username` and `-remoteWrite.basicAuth.password` command-line flags. See `./vmagent -help` output for these flags:

```
  -remoteWrite.basicAuth.password array
        Optional basic auth password to use for -remoteWrite.url. If multiple args are set, then they are applied independently for the corresponding -remoteWrite.url
        Supports array of values separated by comma or specified via multiple flags.
  -remoteWrite.basicAuth.username array
        Optional basic auth username to use for -remoteWrite.url. If multiple args are set, then they are applied independently for the corresponding -remoteWrite.url
        Supports array of values separated by comma or specified via multiple flags.
```

If you still stick with passing basic auth creds via `-remoteWrite.url`, then try `vmagent` built from the commit ab53cb6f7b11101f4acd9ad762348601bfb4f063 . This commit will be included in the next release if it works as expected. See [build instructions for vmagent](https://victoriametrics.github.io/vmagent.html#how-to-build-from-sources).

**Comment by valyala** [comment]
FYI, all the occurrences of `-remoteWrite.url` in logs and metrics are substituted by `secret-url` text by default starting from [v1.41.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). If you need `-remoteWrite.url` in logs and metrics, then pass `-remoteWrite.showURL` command-line flag to `vmagent`.

Closing the issue as resolved. @ofen , feel free re-opening it if you still see plain `-remoteWrite.url` in logs / metrics.

**Comment by Allineer** [comment]
@valyala, why not only username and password are masked?
URL are useful and required in metrics and logs, but without credentials.

**Comment by ofen** [comment]
Looks like start-up logs still contains value of **remoteWrite.url** flag with unmasked username and password
```
...
2020-09-24T19:17:40.165Z  info  VictoriaMetrics/lib/logger/flag.go:21  flag "remoteWrite.tlsKeyFile" = "secret"
2020-09-24T19:17:40.165Z  info  VictoriaMetrics/lib/logger/flag.go:21  flag "remoteWrite.tlsServerName" = ""
2020-09-24T19:17:40.165Z  info  VictoriaMetrics/lib/logger/flag.go:21  flag "remoteWrite.tmpDataPath" = "vmagent-remotewrite-data"
2020-09-24T19:17:40.165Z  info  VictoriaMetrics/lib/logger/flag.go:21  flag "remoteWrite.url" = "https://my_username:my_password@monitoring.local/api/v1/write"
2020-09-24T19:17:40.165Z  info  VictoriaMetrics/lib/logger/flag.go:21  flag "remoteWrite.urlRelabelConfig" = ""
...
```

**Comment by valyala** [comment]
> @valyala, why not only username and password are masked? URL are useful and required in metrics and logs, but without credentials.

Because `-remoteWrite.url` may contain auth tokens in other parts of url. For example, `https://victoriametrics.host/<secret_auth_token_here>/api/v1/write` . It is impossible to mask parts of url with secret information in general case. So the best solution is just hide the entire url. If you want seeing the `-remoteWrite.url` in logs and metrics while using basic auth, then just specify `-remoteWrite.basicAuth.username` and `-remoteWrite.basicAuth.password` values separately from `-remoteWrite.url` and pass `-remoteWrite.showURL` command-line flag when starting `vmagent`. For example:

```
./vmagent -remoteWrite.url=http://localhost:8428/api/v1/write -remoteWrite.basicAuth.username=foo -remoteWrite.basicAuth.password=bar -remoteWrite.showURL
```

> Looks like start-up logs still contains value of remoteWrite.url flag with unmasked username and password

Good catch! It should be fixed in the commit 127537d631c75d4169254941d6b69ce6cfd9864d

**Comment by valyala** [comment]
VictoriaMetrics shouldn't show `-remoteWrite.url` in logs and in metrics starting from [v1.42.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue as done.
